service/history/hsm/tasks.go

```diff
@@ -59,19 +55,11 @@ type Task interface {
 	// If a destination is set, a task will be scheduled on the outbound queue.
 	// Currently Destination and Deadline are mutually exclusive.
 	Destination() string
-	// Whether this task is automatically invalidated if the machine transitioned since generating it.
-	Concurrent() bool
-}
-
-// A concurrent task can run concurrently with other tasks.
-type ConcurrentTask interface {
-	Task
-	// Validate checks if the [ConcurrentTask] is still valid for processing in
-	// either active or standby queue task executor.
-	// Must return ErrStaleReference if the task is no longer valid.
-	// A typical implementation may check if the state of the machine is still
-	// relevant for running this task.
-	Validate(node *Node) error
+	// Validate checks if the task is still valid for processing for the current node state.
+	// Implementors may return [ErrStaleReference] or [consts.ErrWorkflowCompleted] if the task is no longer valid.
+	// A typical implementation may use [node.CheckRunning], [ValidateNotTransitioned], or check if the state of the
+	// machine is still relevant for running this task.
+	Validate(ref *persistencespb.StateMachineRef, node *Node) error
```

Comment:
We are pushing the validate logic to component author anyway, so why not let the task executor perform the validation? Framework still provides util func like ValidateNotTransitioned the make the implementation easy. 

I can see it has some benefits like separate the logic, explicitly asking author to think about dedup logic, etc. But at the same time also adds complexity to the framework.

Reply:
I found this to be convenient but I can see your point where we might want to put this `Validate` method on a more static object where we can inject config and other dependencies.

Let's keep this for now because it's already an improvement. We're designing a longer term solution ATM anyways.